libs/community/langchain_community/chat_message_histories/mssql.py

```diff
@@ -0,0 +1,234 @@
+        with connection.cursor() as cursor:
+            for query in queries:
+                cursor.execute(query)
+                cursor.commit()
+
+    @staticmethod
+    async def acreate_tables(
+        connection: aioodbc.Connection, table_name: str, /
+    ) -> None:
+        """Create the table schema in the database and create relevant indexes."""
+        queries = _create_table_and_index(table_name)
+        cursor = await connection.cursor()
+        for query in queries:
+            await cursor.execute(query)
+            await cursor.commit()
+        await cursor.close()
+
+    @staticmethod
+    def drop_table(connection: pyodbc.Connection, table_name: str, /) -> None:
+        """Delete the table schema in the database.
+
+        WARNING:
+          This will delete the given table from the database including all the databases in the table and the schema of the table.
+
+        Args:
+          connection: The database engine.
+          table_name: the name of the table to drop
+        """
+
+        query = _delete_table_query(table_name)
+        with connection.cursor() as cursor:
+            cursor.execute(query)
+            cursor.commit()
+
+    @staticmethod
+    async def adrop_table(connection: aioodbc.Connection, table_name: str, /) -> None:
+        """Delete the table schema in the database.
+
+        WARNING:
+          This will delete the given table from the database including all the databases in the table and the schema of the table.
+
+        Args:
+          connection: The database engine.
+          table_name: the name of the table to drop
+        """
+        query = _delete_table_query(table_name)
+        cursor = await connection.cursor()
+        await cursor.execute(query)
+        await cursor.commit()
+        await cursor.close()
+
+    def add_messages(self, messages: Sequence[BaseMessage]) -> None:
+        """Add messages to the chat message history."""
+        if self._connection is None:
+            raise ValueError("No connection to the database.")
+        values = [
+            (self._session_id, json.dumps(message_to_dict(message)))
+            for message in messages
+        ]
+        query = _insert_message_query(self._table_name)
+        with self._connection.cursor() as cursor:
+            cursor.executemany(query, values)
+            self._connection.commit()
+
+    async def aadd_messages(self, messages: Sequence[BaseMessage]) -> None:
+        """Add messages to the chat message history."""
+        if self._aconnection is None:
+            raise ValueError("No connection to the database.")
+
+        values = [
+            (self._session_id, json.dumps(message_to_dict(message)))
+            for message in messages
+        ]
+        query = _insert_message_query(self._table_name)
+        cursor = await self._aconnection.cursor()
+        await cursor.executemany(query, values)
+        await cursor.commit()
+        await cursor.close()
```

Comment:
Could you create a context manager to handle the closing of the cursor

Reply:
I must have really missed something. When I was implementing it, I got an error saying that context managers were not implemented for aioodbc and just went on my way. But according to the front page of their repo, they do https://github.com/aio-libs/aioodbc

Will implement and push up